src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/HubMessage.cs

```diff
@@ -1,12 +1,45 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System.Collections.Generic;
+
 namespace Microsoft.AspNetCore.SignalR.Internal.Protocol
 {
     public abstract class HubMessage
     {
         protected HubMessage()
         {
         }
+
+        // Initialize with capacity 4 for the 2 built in protocols and 2 data encoders
+        private readonly List<SerializedMessage> _serializedMessages = new List<SerializedMessage>(4);
+
+        public byte[] WriteMessage(HubProtocolReaderWriter protocolReaderWriter)
+        {
+            for (var i = 0; i < _serializedMessages.Count; i++)
+            {
+                if (_serializedMessages[i].ProtocolReaderWriter.Equals(protocolReaderWriter))
+                {
+                    return _serializedMessages[i].Message;
+                }
+            }
+
+            var bytes = protocolReaderWriter.WriteMessage(this);
+            _serializedMessages.Add(new SerializedMessage(protocolReaderWriter, bytes));
```

Comment:
If we capped the cache size, that would relieve some of my concerns about client-controlled linear memory growth from the cache.

Reply:
I don’t really understand the concern. Are you worrried about a misconfigured application? We can cap it at 10 if that makes you feel better.